### PR TITLE
upgrade file store implementation

### PR DESCRIPTION
**Pull request by povolev15**
implemented the upgrade file

issues #7353

### CODE REVIEW

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/FileServiceImpl.java [hunk 1]
```diff
@@ -24,6 +24,7 @@
 /** Standard implementation of the {@link FileService} {@link com.hedera.node.app.spi.Service}. */
 public final class FileServiceImpl implements FileService {
     public static final String BLOBS_KEY = "FILES";
+    public static final String UPGRADE_FILE_KEY = "UPGRADE_FILE";
```
Comment (kimbor) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/FileServiceImpl.java:27]: This will replace UPGRADE_FILES_KEY in FreezeServiceImpl. I will do a follow-up PR to delete it from there.

**Review by kimbor — COMMENTED** [review]
Some comments:
1) We should standardize on either UpdateFile or UpgradeFile. You're mostly calling them Upgrade here, I mostly called them Update over in networkAdmin. In the protobuf, Freeze transaction calls it update_file so IMO update is better but I could be persuaded. Upgrade is less generic.
2) Are there concerns about causing a network lag during serialization given the size of the upgrade file? Would it be better to spawn a new thread to write the file to the merkle tree? Maybe Richard has already thought of this when designing the stores but I wanted to raise it just in case.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/FileServiceImpl.java (hunk 1 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/FileServiceImpl.java:27]: ok no problem

File: hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/FilesConfig.java [hunk 2]
```diff
@@ -30,4 +30,5 @@ public record FilesConfig(
         // @ConfigProperty(defaultValue = "150-159") Pair<Long, Long> softwareUpdateRange,
         @ConfigProperty(defaultValue = "123") long throttleDefinitions,
         @ConfigProperty(defaultValue = "1000000") long maxNumber,
-        @ConfigProperty(defaultValue = "1024") int maxSizeKb) {}
+        @ConfigProperty(defaultValue = "1024") int maxSizeKb,
+        @ConfigProperty(defaultValue = "150") long upgradeFileNumber) {}
```
[the remark this answers is not in the document]
Comment (kimbor) [reply hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/FilesConfig.java:34]: See VersionConfig.java. It uses the record SemanticVersion as a type.

[the remark this answers is not in the document]
Comment (povolev15) [reply hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/FilesConfig.java:34]: Given we have only a number for "System Entities", it should be just a number here, and FileService will convert to an ID for the current node/shard, which the update file always will be for the node, not a particular ledger.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileUpdateHandler.java [hunk 3]
```diff
@@ -104,6 +118,21 @@ public void handle(@NonNull final HandleContext handleContext) throws HandleExce
...
+    private void handleUpdateUpgradeFile(FileUpdateTransactionBody fileUpdate, HandleContext handleContext) {
+        final var fileStore = handleContext.writableStore(WritableUpgradeStore.class);
+        // empty old upgrade file
+        while (fileStore.poll() != null)
+            ;
+        final var file = new File.Builder()
+                .fileId(fileUpdate.fileIDOrThrow())
+                .contents(fileUpdate.contents())
+                .deleted(false)
+                .expirationTime(fileUpdate.expirationTimeOrElse(EXPIRE_NEVER).seconds())
```
Comment (kimbor) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileUpdateHandler.java:130]: Do we really want upgrade files to never expire? Seems like setting them to expire in a week or a month would be more than enough

**Comment by kimbor** [comment]
> 1. We should standardize on either UpdateFile or UpgradeFile. You're mostly calling them Upgrade here, I mostly called them Update over in networkAdmin. In the protobuf, Freeze transaction calls it update_file so IMO update is better but I could be persuaded. Upgrade is less generic.

Per discussion, we will standardize on UpgradeFile. I will do a separate PR to rename things over in networkAdmin (with the exception of FreezeTransaction fields in the protobuf which can't easily be renamed) so that we're consistent.

> 2. Are there concerns about causing a network lag during serialization given the size of the upgrade file? Would it be better to spawn a new thread to write the file to the merkle tree? Maybe Richard has already thought of this when designing the stores but I wanted to raise it just in case.

Per discussion, no concerns here.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileGetInfoHandler.java [hunk 4]
```diff
@@ -101,8 +107,26 @@ public void validate(@NonNull final QueryContext context) throws PreCheckExcepti
...
             @NonNull final FileID fileID,
             @NonNull final ReadableFileStore fileStore,
-            @NonNull final LedgerConfig ledgerConfig) {
-        final var meta = fileStore.getFileMetadata(fileID);
+            @NonNull final LedgerConfig ledgerConfig,
+            @NonNull final ReadableUpgradeStoreImpl upgradeStore,
+            @NonNull final FilesConfig fileServiceConfig) {
+
+        FileMetadata meta = null;
+        if (fileID.fileNum() == fileServiceConfig.upgradeFileNumber()) {
```
[the remark this answers is not in the document]
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileGetInfoHandler.java:115]: done thanks

**Review by kimbor — APPROVED** [review]
lgtm!

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/ReadableUpgradeStoreImpl.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        return upgradeState.peek();
+    }
+
+    @NonNull
+    public Iterator<File> iterator() {
+        return upgradeState.iterator();
+    }
+
+    @NonNull
+    public Bytes getFull() throws IOException {
```
Comment (tinker-michaelj) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/ReadableUpgradeStoreImpl.java:68]: This method provides a higher-level abstraction than is available on the `QueueState`, and makes a lot of sense in a `Store`.

Do we really need the other methods that just delegate to a `QueueState` method? 🤔

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/WritableUpgradeStore.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+     * Create a new {@link WritableUpgradeStore} instance.
+     *
+     * @param states The state to use.
+     */
+    public WritableUpgradeStore(@NonNull final WritableStates states) {
+        super(states);
+        this.writableUpgradeState = requireNonNull(states.getQueue(getStateKey()));
+    }
+
+    public void add(@NonNull File file) {
```
Comment (tinker-michaelj) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/WritableUpgradeStore.java:48]: I'd have expected the signature of this method to be,
```
/**
 * Given the next part of the upgrade file, appends it to the parts already stored in state.
*/
public void append(@NonNull Bytes nextFilePart) {
```
...even if behind the scenes we wrap these `Bytes` in a `File` message to get convenient re-use of the `FIle.PROTOBUF` codec.

As written, imo this is a bit confusing.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/WritableUpgradeStore.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        requireNonNull(file);
+        writableUpgradeState.add(file);
+    }
+
+    @Nullable
+    public File poll() {
```
Comment (tinker-michaelj) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/WritableUpgradeStore.java:54]: Could we replace this with a,
```
public void resetFileContents() {
```
...method to provide a higher-level API to the handler?

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileAppendHandler.java [hunk 8]
```diff
@@ -76,21 +78,28 @@ public void preHandle(@NonNull final PreHandleContext context) throws PreCheckEx
...
         if (data == null || data.length() <= 0) {
             logger.debug("FileAppend: No data to append");
         }
 
         if (target == null) {
             throw new HandleException(INVALID_FILE_ID);
         }
-        final var fileStore = context.writableStore(WritableFileStoreImpl.class);
+
+        // the update file always will be for the node, not a particular ledger that's why we just compare th num
```
Comment (tinker-michaelj) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileAppendHandler.java:96]: Hmm it's not that important until we have multiple shards deployed, but I think we _do_ want to validate the entire `FileID` here, not only the number.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileAppendHandler.java [hunk 9]
```diff
@@ -122,4 +131,20 @@ public void handle(@NonNull final HandleContext context) throws HandleException
...
         fileStore.put(fileBuilder.build());
     }
+
+    private void handleAppendUpgradeFile(FileAppendTransactionBody fileAppend, HandleContext handleContext) {
+        final var fileStore = handleContext.writableStore(WritableUpgradeStore.class);
+        File file = fileStore.peek();
+        if (file == null) {
+            throw new HandleException(INVALID_FILE_ID);
+        }
+        final var appendedFile = new File.Builder()
```
Comment (tinker-michaelj) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileAppendHandler.java:141]: I don't think it makes sense to store the `0.0.150` metadata in **each `File` part inside the `FCQueue`.**

Imo we should have a normal `KVState` entry for the `0.0.150` metadata; and then just keep its _data_ in the `WritableUpgradeStore`.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileGetInfoHandler.java [hunk 10]
```diff
@@ -101,8 +107,27 @@ public void validate(@NonNull final QueryContext context) throws PreCheckExcepti
...
+        // the update file always will be for the node, not a particular ledger that's why we just compare th num
+        if (fileID.fileNum() == fileServiceConfig.upgradeFileNumber()) {
+            final var file = upgradeStore.peek();
+            if (file != null) {
+                meta = new FileMetadata(
+                        file.fileId(),
+                        Timestamp.newBuilder().seconds(file.expirationTime()).build(),
+                        file.keys(),
+                        file.contents(),
+                        file.memo(),
```
Comment (tinker-michaelj) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileGetInfoHandler.java:124]: As noted, I think we should be getting this metadata from the normal `KVStore`.

There's also an important detail about the `memo` field we return for file `0.0.150` in this query.

It is not an actual memo, but instead the SHA-384 hash of the current contents (see [here](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/context/primitives/StateView.java#L386) in `mono-service`.)

**Review by tinker-michaelj — CHANGES_REQUESTED** [review]
Great start, functionality all looks correct except the `memo` field in `getFileInfo()` (as noted, should be the SHA-384 hash of the current `0.0.150` contents).

But I think we definitely want to keep the `0.0.150` metadata in the same `KVState` as all the "user-space" file metadata, instead of repeating it in every `File` part appended to the `QueueState`.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/WritableUpgradeStore.java (hunk 7 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/WritableUpgradeStore.java:54]: fixed.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileGetInfoHandler.java (hunk 10 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileGetInfoHandler.java:124]: fixed thanks

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/ReadableUpgradeStoreImpl.java (hunk 5 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/ReadableUpgradeStoreImpl.java:68]: Yes, to implement the interface. WritableQueueState interface

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/WritableUpgradeStore.java (hunk 6 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/WritableUpgradeStore.java:48]: We are storing a queue of files and not a queue of bytes. If we store a queue of Bytes, it adds a lot of complexity because anyone who accesses it needs to read both storages.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileAppendHandler.java (hunk 9 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileAppendHandler.java:141]: It is relevant to the top comment I mentioned. It adds a lot of complexity, as I mentioned above. It will make it hard to understand that it is stored in two different places.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileAppendHandler.java (hunk 8 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileAppendHandler.java:96]: We will talk about it during our daily call.

**Comment by kimbor** [comment]
I created issue #7568 to remove the dead code from networkadmin that is replaced by code in this PR

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileAppendHandler.java (hunk 8 above)
Comment (kimbor) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileAppendHandler.java:96]: Per modularization call on Tuesday, upgrade file should apply to the whole network. It should get clamped shard=0, realm=0

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileUpdateHandler.java [hunk 11]
```diff
@@ -82,20 +82,27 @@ public void handle(@NonNull final HandleContext handleContext) throws HandleExce
...
 
-        // the update file always will be for the node, not a particular ledger that's why we just compare th num
+        // the update file always will be for the node, not a particular ledger that's why we just compare the fileNum
+        // and ignore shard and realm
         if (fileUpdate.fileIDOrThrow().fileNum() == fileServiceConfig.upgradeFileNumber()) {
             handleUpdateUpgradeFile(fileUpdate, handleContext);
             return;
         }
 
+        // copy over just the fileNum from the update file
```
Comment (jsync-swirlds) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileUpdateHandler.java:99]: This isn't an update file, at this point the code is working with a normal file.
Upgrade file returns on line 95.

File: hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/FileTestBase.java [hunk 12]
```diff
@@ -77,12 +77,12 @@ public class FileTestBase {
     protected final FileID WELL_KNOWN_FILE_ID =
             FileID.newBuilder().fileNum(1_234L).build();
     protected final FileID WELL_KNOWN_UPGRADE_FILE_ID =
-            FileID.newBuilder().fileNum(150L).build();
+            FileID.newBuilder().fileNum(150L).shardNum(50L).realmNum(50L).build();
     protected final FileID WELL_KNOWN_SYSTEM_FILE_ID =
-            FileID.newBuilder().fileNum(150L).build();
+            FileID.newBuilder().fileNum(150L).shardNum(50L).realmNum(50L).build();
```
Comment (jsync-swirlds) [inline hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/FileTestBase.java:82]: System file should probably be different from upgrade file, perhaps 0.0.122

File: hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileAppendTest.java [hunk 13]
```diff
@@ -183,15 +175,6 @@ public static void assertFailsWith(final ResponseCodeEnum status, final Runnable
         assertEquals(status, ex.getStatus());
     }
 
-    private TransactionBody txnWith(final FileAppendTransactionBody op) {
```
Comment (jsync-swirlds) [inline hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileAppendTest.java:184]: Why was this removed?

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileUpdateHandler.java [hunk 14]
```diff
@@ -82,20 +82,27 @@ public void handle(@NonNull final HandleContext handleContext) throws HandleExce
 
         final var fileStore = handleContext.writableStore(WritableFileStore.class);
         final var fileUpdate = handleContext.body().fileUpdateOrThrow();
-        final var maybeFile = fileStore.get(fileUpdate.fileIDOrElse(FileID.DEFAULT));
```
Comment (jsync-swirlds) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileUpdateHandler.java:80]: This is used for the regular files, not upgrade file, so perhaps it should be retained?

Comment (kimbor) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileUpdateHandler.java:80]: It was moved further down in the same method

File: hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileAppendTest.java (hunk 13 above)
Comment (kimbor) [reply hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileAppendTest.java:184]: It was not being called anywhere

Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileAppendTest.java:184]: it can be removed Joseph I think it is from the really first version

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileUpdateHandler.java (hunk 14 above)
Comment (jsync-swirlds) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileUpdateHandler.java:80]: That moved version clamps causes incorrect behavior for normal files, this version is correct.

Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileUpdateHandler.java:80]: fixed

File: hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/FileTestBase.java (hunk 12 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/FileTestBase.java]: I reverted it to 0.0.150

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileUpdateHandler.java (hunk 11 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileUpdateHandler.java]: fixed

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileUpdateHandler.java [hunk 15]
```diff
@@ -104,7 +120,23 @@ public void handle(@NonNull final HandleContext handleContext) throws HandleExce
...
     }
 
-    private void resolveMutableBuilderAttributes(
+    private static void handleUpdateUpgradeFile(FileUpdateTransactionBody fileUpdate, HandleContext handleContext) {
+        final var fileStore = handleContext.writableStore(WritableUpgradeStore.class);
+        // empty old upgrade file
+        fileStore.resetFileContents();
+        final var file = new File.Builder()
+                .fileId(FileID.newBuilder()
+                        .fileNum(fileUpdate.fileIDOrElse(FileID.DEFAULT).fileNum())
```
Comment (jsync-swirlds) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileUpdateHandler.java:129]: This can be `fileIDOrThrow`, or `fileServiceConfig.upgradeFileNumber()`. The orElse for default is incorrect (default file has the wrong file number), and also should never happen, as we've already acquired the fileID correctly before calling this method.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileAppendHandler.java [hunk 16]
```diff
@@ -122,4 +132,25 @@ public void handle(@NonNull final HandleContext context) throws HandleException
         It will not be committed to state until commit is called on the state.--- */
         fileStore.put(fileBuilder.build());
     }
+
+    private static void handleAppendUpgradeFile(FileAppendTransactionBody fileAppend, HandleContext handleContext) {
```
Comment (jsync-swirlds) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileAppendHandler.java:136]: Minor note, making every private method static makes unit tests harder, particularly if you need to spy the class to force an exception.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/codec/BytesCodec.java [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    @Override
+    public boolean fastEquals(@NonNull Bytes item, @NonNull ReadableSequentialData input) throws IOException {
+        throw new UnsupportedOperationException("Cannot be implemented");
+        //        final ReadableSequentialData comparisonView = input.view(measure(input));
+        //        final Bytes other = parseStrict(comparisonView);
+        //        return item.equals(other);
+    }
+}
```
Comment (jsync-swirlds) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/codec/BytesCodec.java:67]: Probably should override the default toBytes, as this particular codec can just return the input instead of trying to serialize the Bytes object then wrapping the result in a new Bytes object.

We should also _(**in a future PR**)_ look into adding the BytesCodec as part of PBJ instead of implementing here.

File: hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/codec/EntityNumCodecTest.java
Comment (jsync-swirlds) [inline hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/codec/EntityNumCodecTest.java:1]: Worth noting that EntityNumCodec is implemented separately in each of Consensus, NetworkAdmin, File, and Token service.  We should _(**in a separate PR**)_ create one unified implementation in SPI and replace the four copies with reference to that one.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/WritableUpgradeStore.java (hunk 6 above)
Comment (jsync-swirlds) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/WritableUpgradeStore.java:55]: After discussion, this will be a SingletonState File and Bytes in the queue state.

File: hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/FileTestBase.java (hunk 12 above)
Comment (jsync-swirlds) [reply hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/FileTestBase.java]: It should be *different*, pick any reasonable non-upgrade file value.  We already have WELL_KNOWN_UPGRADE_FILE_ID, this should not be the same.

Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/FileTestBase.java]: fixed to 122

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/WritableUpgradeStore.java [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    public void add(@NonNull File file) {
+        requireNonNull(file);
+        writableUpgradeState.add(file.contents());
+        writableUpgradeFileState.put(file);
+    }
+
+    public void resetFileContents() {
+        while (writableUpgradeState.removeIf(TRUE_PREDICATE) != null)
```
Comment (tinker-michaelj) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/WritableUpgradeStore.java:62]: This doesn't need to be inside a `while` loop any more, right?

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/codec/BytesCodec.java [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+package com.hedera.node.app.service.file.impl.codec;
+
+import com.hedera.pbj.runtime.Codec;
+import com.hedera.pbj.runtime.io.ReadableSequentialData;
+import com.hedera.pbj.runtime.io.WritableSequentialData;
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.io.IOException;
+
+public class BytesCodec implements Codec<Bytes> {
```
Comment (tinker-michaelj) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/codec/BytesCodec.java:26]: Shouldn't we be using `Bytes.PROTOBUF` instead of implementing this here?

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileAppendHandler.java [hunk 20]
```diff
@@ -122,4 +132,25 @@ public void handle(@NonNull final HandleContext context) throws HandleException
...
+                // this ensures that shard and realm are clamped to zero
+                .fileId(FileID.newBuilder()
+                        .fileNum(fileAppend.fileIDOrThrow().fileNum())
+                        .build())
+                .contents(fileAppend.contents())
+                .deleted(false)
+                .expirationTime(file.expirationTime())
+                .memo(file.memo())
+                .build();
+        fileStore.add(appendedFile);
```
Comment (tinker-michaelj) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileAppendHandler.java:154]: _Nit_ - The_`FileAppend` semantics affect _only_ the file's content, not its metadata.

So I think there should be a `fileStore.append(Bytes extraContents)` method we use here; not an `add()` that takes an entire `File` object.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileAppendHandler.java (hunk 16 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileAppendHandler.java:136]: by mistake sorry I thought it was used also out of the class. fixed it

File: hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/codec/EntityNumCodecTest.java
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/codec/EntityNumCodecTest.java:1]: I'm nut using it anymore I think I can delete it

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/WritableUpgradeStore.java (hunk 18 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/WritableUpgradeStore.java:62]: I'm not sure if I get it right, but we clear the queue before we add a new file during file upgrade

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/codec/BytesCodec.java (hunk 19 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/codec/BytesCodec.java:26]: I didn't find one and I checked what the method is expecting and created one

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileAppendHandler.java (hunk 20 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileAppendHandler.java:154]: The file add is doing excatly it inside.
```
    public void add(@NonNull File file) {
        requireNonNull(file);
        writableUpgradeState.add(file.contents());
        writableUpgradeFileState.put(file);
    }
```

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/codec/BytesCodec.java (hunk 19 above)
Comment (jsync-swirlds) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/codec/BytesCodec.java:26]: Bytes isn't a generated record (it's part of PBJ runtime), and does not have a .PROTOBUF member.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileGetInfoHandler.java [hunk 21]
```diff
@@ -98,18 +114,44 @@ public void validate(@NonNull final QueryContext context) throws PreCheckExcepti
...
-            @NonNull final LedgerConfig ledgerConfig) {
-        final var meta = fileStore.getFileMetadata(fileID);
+            @NonNull final LedgerConfig ledgerConfig,
+            @NonNull final ReadableUpgradeStoreImpl upgradeStore,
+            @NonNull final FilesConfig fileServiceConfig)
+            throws IOException {
+
+        FileMetadata meta = null;
+        long contentSize = 0L;
+        // the update file always will be for the node, not a particular ledger that's why we just compare the num
```
Comment (jsync-swirlds) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileGetInfoHandler.java:127]: Nit:
upgrade is for the entire network, not a node.  It's across shards and realms, however, which is why we ignore the shard and realm values.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileAppendHandler.java (hunk 20 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileAppendHandler.java:154]: I fixed it as you asked for the appendHandler inside the store

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileGetInfoHandler.java (hunk 21 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileGetInfoHandler.java]: fixed the comment. thank you

**Review by jsync-swirlds — APPROVED** [review]

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/WritableUpgradeStore.java (hunk 18 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/WritableUpgradeStore.java:67]: Oh n/m the semantics of this `removeIf()` are different than the `j.u.Collection.removeIf()`. 😝

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/codec/BytesCodec.java (hunk 19 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/codec/BytesCodec.java:26]: 😢 ...ah got it, I didn't mean we had to switch away from using `File` as a wrapper for each "chunk" of content, but this works too. 👍

**Review by tinker-michaelj — APPROVED** [review]
LGTM, thanks @povolev15 and @jsync-swirlds 🥂
